edgraph/access_ee_test.go

```diff
@@ -0,0 +1,180 @@
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+var (
+	secretkey string = "secretkey"
+
+	userDataList = []userData{
+		userData{1234567890, "user1", []string{"701", "702"}},
+		userData{2345678901, "user2", []string{"703", "701"}},
+		userData{3456789012, "user3", []string{"702", "703"}},
+	}
+)
+
+func generateJWT(namespace uint64, userId string, groupIds []string, exp int64) (string, error) {
+	var e1 int64
+	var token *jwt.Token
+
+	if exp == 0 {
+		e1 = time.Now().Add(time.Minute * 30).Unix()
+	} else {
+		e1 = exp
+	}
+
+	if groupIds != nil {
+		token = jwt.NewWithClaims(jwt.SigningMethodHS256, &jwt.MapClaims{
+			"namespace": namespace,
+			"userid":    userId,
+			"groups":    groupIds,
+			"exp":       e1,
+		})
+	} else {
+		token = jwt.NewWithClaims(jwt.SigningMethodHS256, &jwt.MapClaims{
+			"namespace": namespace,
+			"userid":    userId,
+			"exp":       e1,
+		})
+	}
+
+	tokenString, err := token.SignedString([]byte(worker.Config.HmacSecret))
+	if err != nil {
+		_ = fmt.Errorf("Something Went Wrong: %s", err.Error())
+		return "", err
+	}
+	return tokenString, nil
+}
+
+func sliceCompare(x, y []string) bool {
+	if len(x) != len(y) {
+		return false
+	}
+
+	for i, s := range x {
+		if s != y[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func TestValidateToken(t *testing.T) {
+	var (
+		err         error
+		tokenString string
+	)
+
+	for _, userdata := range userDataList {
+		if tokenString, err = generateJWT(userdata.namespace, userdata.userId, userdata.groupIds, 0); err != nil {
+			t.Errorf("JWT token string generation error: %s", err)
+			return
+		}
+
+		ud, err := validateToken(tokenString)
+		if err != nil {
+			t.Errorf("validateToken() error: %s", err)
+			return
+		}
+		if ud.namespace != userdata.namespace || ud.userId != userdata.userId || !sliceCompare(ud.groupIds, userdata.groupIds) {
+			t.Errorf("Actual output %+v is not equal to the expected output %+v", userdata, ud)
+		}
+	}
+}
+
+func TestGetAccessJwt(t *testing.T) {
+	var (
+		err    error
+		tokstr string
+		jwtstr string
+		exp    int64 = time.Now().Add(worker.Config.AccessJwtTtl).Unix()
+	)
+
+	grpLst := []acl.Group{
+		{
+			Uid:     "100",
+			GroupID: "1001",
+			Users:   []acl.User{},
+			Rules:   []acl.Acl{},
+		},
+		{
+			Uid:     "101",
+			GroupID: "1011",
+			Users:   []acl.User{},
+			Rules:   []acl.Acl{},
+		},
+		{
+			Uid:     "102",
+			GroupID: "1021",
+			Users:   []acl.User{},
+			Rules:   []acl.Acl{},
+		},
+	}
+
+	for _, userdata := range userDataList {
+		g := acl.GetGroupIDs(grpLst)
+		if tokstr, err = generateJWT(userdata.namespace, userdata.userId, g, exp); err != nil {
+			t.Errorf("JWT token string generation error: %s", err)
+			return
+		}
+		t1 := strings.Split(tokstr, ".")
+
+		jwtstr, err = getAccessJwt(userdata.userId, grpLst, userdata.namespace)
+		if err != nil {
+			t.Errorf("getAccessJwt () error: %s", err)
+			return
+		}
+		t2 := strings.Split(jwtstr, ".")
+
+		/*
+		 * JWT token format = [header].[payload].[signature].
```

Comment:
We can use a helper function to do this validation and use it in the two functions below.